<a href="https://colab.research.google.com/github/ElyorS/AI-application-system/blob/main/12204556_week7_hw_RNN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

***Task 2.1***
========

In [1]:
import tensorflow as tf

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images, test_images = train_images / 255.0, test_images / 255.0

# Initialize the RNN model with Dropout and Batch Normalization
model = tf.keras.Sequential([
    tf.keras.layers.SimpleRNN(128, activation='relu', return_sequences=True, input_shape=(28, 28)),
    tf.keras.layers.BatchNormalization(),  # Add Batch Normalization
    tf.keras.layers.SimpleRNN(128, activation='relu'),
    tf.keras.layers.BatchNormalization(),  # Add Batch Normalization
    tf.keras.layers.Dense(10, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(train_images, train_labels, epochs=5)

# Evaluate the model
test_loss, test_acc = model.evaluate(test_images, test_labels)
print("Test accuracy:", test_acc)


11490434/11490434 [==============================] - 1s 0us/step
Epoch 1/5
1875/1875 [==============================] - 62s 30ms/step - loss: 0.3780 - accuracy: 0.8796
Epoch 2/5
1875/1875 [==============================] - 46s 25ms/step - loss: 0.1526 - accuracy: 0.9550
Epoch 3/5
1875/1875 [==============================] - 49s 26ms/step - loss: 0.1312 - accuracy: 0.9629
Epoch 4/5
1875/1875 [==============================] - 45s 24ms/step - loss: 0.1131 - accuracy: 0.9678
Epoch 5/5
313/313 [==============================] - 3s 7ms/step - loss: 0.1074 - accuracy: 0.9691
Test accuracy: 0.9690999984741211


This code adds Batch Normalization layers to the RNN model to improve generalization and training speed. You can also add Dropout layers if you want to experiment with further regularization.

Test accuracy: 0.9690999984741211

***Task 2.2***
==============

In [2]:
import tensorflow as tf

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images, test_images = train_images / 255.0, test_images / 255.0

# Define a list of activation functions to experiment with
activation_functions = ['relu', 'tanh', 'sigmoid']

for activation_function in activation_functions:
    # Initialize the RNN model with the specified activation function
    model = tf.keras.Sequential([
        tf.keras.layers.SimpleRNN(128, activation=activation_function, return_sequences=True, input_shape=(28, 28)),
        tf.keras.layers.SimpleRNN(128, activation=activation_function),
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    # Compile the model
    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

    # Train the model
    model.fit(train_images, train_labels, epochs=5)

    # Evaluate the model
    test_loss, test_acc = model.evaluate(test_images, test_labels)

    print(f"Activation Function: {activation_function}")
    print("Test accuracy:", test_acc)
    print("------------------")


Epoch 1/5
1875/1875 [==============================] - 40s 21ms/step - loss: 0.3433 - accuracy: 0.8916
Epoch 2/5
1875/1875 [==============================] - 37s 20ms/step - loss: 0.1502 - accuracy: 0.9586
Epoch 3/5
1875/1875 [==============================] - 38s 21ms/step - loss: 0.1231 - accuracy: 0.9660
Epoch 4/5
1875/1875 [==============================] - 38s 20ms/step - loss: 0.1065 - accuracy: 0.9707
Epoch 5/5
313/313 [==============================] - 2s 7ms/step - loss: 0.0951 - accuracy: 0.9725
Activation Function: relu
Test accuracy: 0.9725000262260437
------------------
Epoch 1/5
1875/1875 [==============================] - 41s 21ms/step - loss: 0.3170 - accuracy: 0.9045
Epoch 2/5
1875/1875 [==============================] - 43s 23ms/step - loss: 0.1912 - accuracy: 0.9446
Epoch 3/5
1875/1875 [==============================] - 39s 21ms/step - loss: 0.1694 - accuracy: 0.9508
Epoch 4/5
1875/1875 [==============================] - 45s 24ms/step - loss: 0.1562 - accuracy: 0.955

Test accuracy with Relu: 0.9725000262260437
________________________
Test accuracy with Tanh: 0.9435999989509583
________________________
Test accuracy with Sigmoid: 0.9225999712944031

In this code, I iterated through three different activation functions (ReLU, tanh, and sigmoid) and train a separate RNN model for each. After training, we evaluate and print the test accuracy for each activation function.
The impact of activation functions on RNN performance can vary, and it's important to experiment with different functions to see which one works best for your specific problem. In. this case, Relu had the best accurasy, followed by Tanh and Sigmoid.

***Task 2.3***
============

In [3]:
import tensorflow as tf

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images, test_images = train_images / 255.0, test_images / 255.0

# Initialize the RNN model with more RNN layers
model = tf.keras.Sequential([
    tf.keras.layers.SimpleRNN(128, activation='relu', return_sequences=True, input_shape=(28, 28)),
    tf.keras.layers.SimpleRNN(128, activation='relu', return_sequences=True),  # Add another RNN layer
    tf.keras.layers.SimpleRNN(128, activation='relu', return_sequences=True),  # Add another RNN layer
    tf.keras.layers.SimpleRNN(128, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(train_images, train_labels, epochs=5)

# Evaluate the model
test_loss, test_acc = model.evaluate(test_images, test_labels)
print("Test accuracy:", test_acc)


Epoch 1/5
1875/1875 [==============================] - 94s 49ms/step - loss: 0.2983 - accuracy: 0.9094
Epoch 2/5
1875/1875 [==============================] - 80s 43ms/step - loss: 0.1382 - accuracy: 0.9620
Epoch 3/5
1875/1875 [==============================] - 82s 44ms/step - loss: 0.1186 - accuracy: 0.9674
Epoch 4/5
1875/1875 [==============================] - 80s 43ms/step - loss: 0.1048 - accuracy: 0.9710
Epoch 5/5
313/313 [==============================] - 5s 13ms/step - loss: 0.1059 - accuracy: 0.9710
Test accuracy: 0.9710000157356262


In this code, I have added two more SimpleRNN layers to the RNN model, making it deeper. This allows the model to potentially recognize more complex sequential patterns. The impact of adding more layers may vary depending on the specific dataset and problem.

Test accuracy: 0.9710000157356262

Task 2.4

In [4]:
import tensorflow as tf

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images, test_images = train_images / 255.0, test_images / 255.0

# Define a list of optimizers to experiment with
optimizers = ['sgd', 'rmsprop', 'adam']

for optimizer in optimizers:
    # Initialize the RNN model
    model = tf.keras.Sequential([
        tf.keras.layers.SimpleRNN(128, activation='relu', return_sequences=True, input_shape=(28, 28)),
        tf.keras.layers.SimpleRNN(128, activation='relu'),
        tf.keras.layers.Dense(10, activation='softmax')
    ])

    # Compile the model with the specified optimizer
    if optimizer == 'sgd':
        optimizer_obj = tf.keras.optimizers.SGD(learning_rate=0.01)
    elif optimizer == 'rmsprop':
        optimizer_obj = tf.keras.optimizers.RMSprop(learning_rate=0.001)
    else:  # Default to Adam
        optimizer_obj = 'adam'

    model.compile(optimizer=optimizer_obj, loss='sparse_categorical_crossentropy', metrics=['accuracy'])

    # Train the model
    model.fit(train_images, train_labels, epochs=5)

    # Evaluate the model
    test_loss, test_acc = model.evaluate(test_images, test_labels)

    print(f"Optimizer: {optimizer}")
    print("Test accuracy:", test_acc)
    print("------------------")


Epoch 1/5
1875/1875 [==============================] - 38s 19ms/step - loss: 0.6743 - accuracy: 0.7746
Epoch 2/5
1875/1875 [==============================] - 36s 19ms/step - loss: 0.1766 - accuracy: 0.9474
Epoch 3/5
1875/1875 [==============================] - 36s 19ms/step - loss: 0.1255 - accuracy: 0.9625
Epoch 4/5
1875/1875 [==============================] - 38s 20ms/step - loss: 0.1026 - accuracy: 0.9692
Epoch 5/5
313/313 [==============================] - 3s 8ms/step - loss: 0.1094 - accuracy: 0.9680
Optimizer: sgd
Test accuracy: 0.9679999947547913
------------------
Epoch 1/5
1875/1875 [==============================] - 38s 19ms/step - loss: 0.3524 - accuracy: 0.8891
Epoch 2/5
1875/1875 [==============================] - 35s 19ms/step - loss: 0.1497 - accuracy: 0.9586
Epoch 3/5
1875/1875 [==============================] - 35s 19ms/step - loss: 0.1215 - accuracy: 0.9682
Epoch 4/5
1875/1875 [==============================] - 37s 20ms/step - loss: 0.1087 - accuracy: 0.9720
Epoch 5/5

Test accuracy with sgd: 0.9679999947547913
__________________
Test accuracy with rmsprop: 0.9743000268936157
__________________
Test accuracy with adam: 0.9746000170707703

In this code, I have used three different optimizers (SGD, RMSprop, and Adam), train a separate RNN model for each optimizer, and evaluate and print the test accuracy.
Different optimizers can significantly affect training speed and performance, so it's important to experiment with various optimizers to find the one that works best for your specific problem. In this case, Adam had the best accuracy followed by rmsprop and sgd.

***Task 2.5***
=============


In [5]:
import tensorflow as tf

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize the images
train_images, test_images = train_images / 255.0, test_images / 255.0

# Initialize the RNN model (you can choose any variant from the previous tasks)
model = tf.keras.Sequential([
    tf.keras.layers.SimpleRNN(128, activation='relu', return_sequences=True, input_shape=(28, 28)),
    tf.keras.layers.SimpleRNN(128, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
])

# Compile the model (you can use any optimizer or configuration as needed)
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Train the model (you can choose the number of epochs as needed)
model.fit(train_images, train_labels, epochs=5)

# Evaluate the model with test data
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f"Test accuracy: {test_acc}")

Epoch 1/5
1875/1875 [==============================] - 50s 26ms/step - loss: 0.3361 - accuracy: 0.8952
Epoch 2/5
1875/1875 [==============================] - 41s 22ms/step - loss: 0.1462 - accuracy: 0.9587
Epoch 3/5
1875/1875 [==============================] - 42s 23ms/step - loss: 0.1199 - accuracy: 0.9661
Epoch 4/5
1875/1875 [==============================] - 43s 23ms/step - loss: 0.1109 - accuracy: 0.9693
Epoch 5/5
313/313 [==============================] - 3s 7ms/step - loss: 0.0886 - accuracy: 0.9736
Test accuracy: 0.9735999703407288


In this code, I loaded the trained RNN model, evaluated it using the test dataset, and printed the test loss and test accuracy.
This task is important to assess how well the RNN model generalizes to unseen data and provides insights into its overall performance.

0.9735999703407288

